prog/hints.go

```diff
@@ -83,10 +123,182 @@ func generateHints(p *Prog, compMap CompMap, c *Call, arg Arg, exec func(newP *P
 	}
 }
 
-func checkConstArg(arg *ConstArg, compMap CompMap, cb func(newArg Arg)) {
-	for v, _ := range compMap[arg.Val] {
-		cb(constArg(arg.typ, v))
+// Creates hints for a ConstArg and calls cb() for each of the obtained
+// programs.
+func checkConstArg(arg *ConstArg, compMap CompMap, cb func(arg, newArg Arg)) {
+	replacersSet := getReplacersForVal(arg.Val, compMap)
+	for newV, _ := range replacersSet {
+		newArg := constArg(arg.typ, newV)
+		cb(arg, newArg)
+	}
+}
+
+// Returns a set of replacers for a given value. A replacer is a value with
+// which an argument's value should be replaced.
+func getReplacersForVal(value uint64, compMap CompMap) uint64Set {
+	replacersSet := make(uint64Set)
+	f := func(transform func(uint64) uint64) {
+		transformedValue := transform(value)
+		originalsSet := make(uint64Set)
+		// Get all different mutations of transformedValue: shrink, expand, ...
+		vals := getMutationsForConstVal(transformedValue)
+		diff := addArrayToSet(originalsSet, vals)
+		// Search for each unseen mutation.
+		for _, v := range diff {
+			compSet, ok := compMap[v]
+			if !ok {
+				continue
+			}
+			for newV, _ := range compSet {
+				// Transform the second operand the same way as the first one.
+				newV = transform(newV)
+				replacersSet[newV] = true
+			}
+		}
+	}
+	// Each transform is related to a different way of encoding an int.
+	transforms := []func(uint64) uint64{
+		// Identity transform.
+		func(v uint64) uint64 { return v },
+		// Revert order of bytes (Big Endian <-> Little Endian) and leave
+		// trailing zero bytes.
+		func(v uint64) uint64 { return reverse(v, false) },
+		// Revert order of bytes (Big Endian <-> Little Endian) and don't leave
+		// trailing zero bytes.
+		func(v uint64) uint64 { return reverse(v, true) },
+	}
+	for _, t := range transforms {
+		f(t)
+	}
+
+	return replacersSet
+}
+
+// Returns an array of different mutations of an ConstArg value. Each of these
+// mutations should be matched against comparison operands.
+func getMutationsForConstVal(v uint64) []uint64 {
+	values := []uint64{v}
+	mutations := []func(uint64) []uint64{
+		shrinkMutation,
+		expandMutation,
+	}
+	for _, m := range mutations {
+		values = append(values, m(v)...)
+	}
+	return values
+}
+
+// Mutation 1: shrink values. Useful in cases like:
+//	void f(int64 v64) {
+//		v32 = (int32) v32;
+//		if (v32 == -1) {...};
+//	}
+// If v64 < 0 and v64's value fits into 32 bits (e.g. v64 = -2), then:
+// uint64(v64) = 0xfffffffffffffffe and uint64(v32) = 0xfffffffe.
+// Thus, the comparison will be (0xfffffffe vs 0xffffffff), and we'll be
```

Comment:
After sign-extension in executor, we will get (0xfffffffffffffffe vs 0xffffffffffffffff), so this first part of the comment does not look right.